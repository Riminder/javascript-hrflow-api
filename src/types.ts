import { ReadStream } from "fs";
export interface HrflowOptions {
  api_secret: string;
  api_user: string;
  webhooks_key?: string;
}

export interface HrflowAPIResponse {
  code: number;
  message: string;
  data?: any;
}

export interface SourcesOptions {
  name?: string;
  page?: number;
  limit?: number;
  sort_by?: string;
  order_by?: string;
} 
export interface profilesSearchingOptions {
  source_keys: Array<string>;
  stage?: string;
  limit?: number;
  page?: number;
  order_by?: string;
  sort_by?: string;
  created_at_min?: Date | number | string;
  created_at_max?: Date | number | string;
  name?: string;
  email?: string;
  location_geopoint?: any;
  location_distance?: number;
  text_keywords?: Array<string>;
  experience_location_geopoint?: {
    lat: number;
    lon: number;
  };
  experience_location_distance?: number;
  experiences_duration_min?: number;
  experiences_duration_max?: number;
  education_location_geopoint?: {
    lat: number;
    lon: number;
  };
  education_location_distance?: number;
  educations_duration_min?: number;
  educations_duration_max?: number;
  skills?: Array<Skills>;
  languages?: Array<Languages>;
  interests?: Array<Interests>;
  tags?: Array<Tags>; 
}

export interface ProfilesScoringOptions {
  source_keys: Array<string>;
  job_key: string;
  board_key: string;
  use_agent?: number;
  stage?: string;
  limit?: number;
  page?: number;
  order_by?: string;
  sort_by?: string;
  created_at_min?: Date | number | string;
  timestamp_end?: Date | number | string;
  name?: string;
  email?: string;
  location_geopoint?: any;
  location_distance?: number;
  text_keywords?: Array<string>;
  experience_location_geopoint?: {
    lat: number;
    lon: number;
  };
  experience_location_distance?: number;
  experiences_duration_min?: number;
  experiences_duration_max?: number;
  education_location_geopoint?: {
    lat: number;
    lon: number;
  };
  education_location_distance?: number;
  educations_duration_min?: number;
  educations_duration_max?: number;
  skills?: Array<Skills>;
  languages?: Array<Languages>;
  interests?: Array<Interests>;
  tags?: Array<Tags>; 
}

export interface JobPostOptions {
  name: string;
  board_key: string;
  agent_key: string;
  url?: string;
  reference?: string;
  summary?: string;
  sections?: Array<Sections>;
  ranges_date?: Array<RangesDate>;
  ranges_float?: Array<RangesFloat>;
  location?: {
    text: string;
    lat: any;
    lng: any;
  };
  skills?: Array<Skills>;
  Interests?: Array<Interests>;
  Languages?: Array<Languages>;
  tags?: Array<Tags>;
  metadatas?: Array<Metadatas>;
  created_at?: number | Date | string;
  members?: Array<string>;
}

export interface jobsScoringOptions {
  board_keys: Array<string>;
  agent_key: string;
  profile_key: string;
  source_key: string;
  text_keywords?: Array<string>;
  tags?: Array<Tags>;
  use_agent?: any;
  page?: number;
  limit?: number;
  order_by?: string;
  sort_by?: string;
  name?: string;
  location_distance?: number;
  location_geopoint?: any;

}
export interface jobsSearchingOptions {
  board_keys: Array<string>;
  page?: number;
  limit?: number;
  order_by?: string;
  sort_by?: string;
  name?: string;
  text_keywords?: Array<string>;
  tags?: Array<Tags>;
  location_distance?: number;
  location_geopoint: any;
}

export interface ProfileOptionId {
  source_key: string;
  key: string;
  profile_email?: string;
  fields?: any;
}

export interface ProfileOptionReference {
  source_key: string;
  profile_reference: string;
  profile_email?: string;
}

export type ProfileOptionIdOrReference = ProfileOptionId | ProfileOptionReference;

export interface jobOptionId {
  job_key: string;
}

export interface jobOptionReference {
  job_reference: string;
}

export type jobOptionIdOrReference = jobOptionId | jobOptionReference;
export interface Metadatas {
  name: string;
  value: any;
}

export interface Labels {
  job_key?: string;
  job_reference?: string;
  stage?: string;
  stage_timestamp?: Date | number;
  rating?: string;
  rating_timestamp?: string;
}
export interface ProfileUpload {
  source_key: string;
  sync_parsing?: number;
  file: ReadStream;
  profile_type?: any;
  profile_content_type?: string;
  profile_reference?: string;
  created_at?: Date | number;
  tags?: Array<Tags>;
  labels?: Labels;
  metadatas?: Array<Metadatas>;
  webhook_parsing_sending?: number;
}

export interface ProfileInfo {
  full_name?: string;
  first_name?: string;
  last_name?: string;
  phone?: string;
  email?: string;
  urls?: {
    from_resume?: Array<string>;
    linkedin?: string;
    twitter?: string;
    facebook?: string;
    github?: string;
    picture?: string;
  };
  location?: {
    text: string
  };
  summary?: string;
}

export interface ConsentAlgoritmic {
  owner: {
      parsing: boolean;
      revealing: boolean;
      embedding: boolean;
      searching: boolean;
      scoring: boolean;
      reasoning: boolean
  };
  controller: {
      parsing: boolean;
      revealing: boolean;
      embedding: boolean;
      searching: boolean;
      scoring: boolean;
      reasoning: boolean
  }
}
export interface ProfileJSON {
  source_key: string;
  info: ProfileInfo;
  profile_reference?: string;
  consent_algoritmic?: ConsentAlgoritmic;
  experiences?: Array<Experience>;
  educations?: Array<Education>;
  educations_duration?: number;
  eperience_duration?: number;
  text?: string;
  skills?: Array<Skills>;
  tags?: Array<Tags>;
  metadatas?: any;
  labels?: Labels;
  languages?: Array<Languages>;
  interests?: Array<Interests>;
  created_at?: number | Date;
}


export interface StagePatchBase {
  source_key: string;
  stage: Stage | null;
}

export interface StagePatchProfileIdFilterId extends StagePatchBase {
  profile_key: string;
  filter_id: string;
}

export interface StagePatchProfileIdFilterReference extends StagePatchBase {
  profile_key: string;
  filter_reference: string;
}

export interface StagePatchProfileReferenceFilterId extends StagePatchBase {
  profile_reference: string;
  filter_id: string;
}

export interface StagePatchProfileReferenceFilterReference extends StagePatchBase {
  profile_reference: string;
  filter_reference: string;
}

export type StagePatch =
  StagePatchProfileIdFilterId |
  StagePatchProfileIdFilterReference |
  StagePatchProfileReferenceFilterId |
  StagePatchProfileReferenceFilterReference;

export interface RatingPatchBase {
  source_key: string;
  rating: number | null;
}

export interface RatingPatchProfileIdFilterId extends RatingPatchBase {
  profile_key: string;
  filter_id: string;
}

export interface RatingPatchProfileIdFilterReference extends RatingPatchBase {
  profile_key: string;
  filter_reference: string;
}

export interface RatingPatchProfileReferenceFilterId extends RatingPatchBase {
  profile_reference: string;
  filter_id: string;
}

export interface RatingPatchProfileReferenceFilterReference extends RatingPatchBase {
  profile_reference: string;
  filter_reference: string;
}

export type RatingPatch =
  RatingPatchProfileIdFilterId |
  RatingPatchProfileIdFilterReference |
  RatingPatchProfileReferenceFilterId |
  RatingPatchProfileReferenceFilterReference;

export interface FilterId {
  filter_id: string;
}

export interface FilterReference {
  filter_reference: string;
}

export type FilterIdOrReference = FilterId | FilterReference;

export interface Experience {
  date_start: string;
  date_end: string;
  title: string;
  company: string;
  location: {
    text: string;
  };
  description: string;
}

export interface Education {
  date_start: string;
  date_end: string;
  title: string;
  school: string;
  location: {
    text: string;
  };
  description: string;
}

export interface Tags {
  name: string;
  value: any
}
export interface Skills {
  name: string;
  value: any
}

export interface Languages {
  name: string;
  value: any
}

export interface Interests {
  name: string;
  value: any
}

export interface Sections {
  name: string;
  title: string;
  description: string;
}

export interface RangesFloat {
  name: string;
  unit: string;
  value_min: number;
  value_max: number;
}

export interface RangesDate {
  name: string;
  value_min?: number | Date | string;
  value_max?:  number | Date | string;
}

export enum Stage {
  NEW = "NEW",
  YES = "YES",
  LATER = "LATER",
  NO = "NO",
}

export enum SortBy {
  DATE_RECEPTION = "date_reception",
  DATE_CREATION = "date_creation",
  LOCATION = "location",
  LOCATION_EXPERIENCE = "location_experience",
  LOCATION_EDUCATION = "location_education",
  SCORE_SEMENTIC = "score_semantic",
  SCORE_PREDECTIVE = "score_predictive",
}

export enum OrderBy {
  DESC = "desc",
  ASC = "asc"
}

export enum Seniority {
  ALL = "all",
  SENIOR = "senior",
  JUNIOR = "junior"
}

export interface WebhooksResponse {
  team_name: string;
  webhook_id: string;
  url: string;
  type: string;
}
