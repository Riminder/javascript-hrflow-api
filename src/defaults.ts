let defaults: any = {
  API_URL: "https://api.hrflow.ai/v1",
  API_SECRET: null,
  api_secret: null,
  api_user: null,
};

export default defaults;