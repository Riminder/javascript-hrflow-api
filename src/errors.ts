import { HrflowAPIResponse } from "./types";

export class APIError extends Error {
  response: HrflowAPIResponse;
  constructor(message: string, response: HrflowAPIResponse)Â {
    super(message);
    this.response = response;
  }
}