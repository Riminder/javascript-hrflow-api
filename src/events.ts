export default [
  "profile.parsing.success",
  "profile.parsing.error",
  "profile.parsing.update",
  "profile.scoring.success",
  "profile.scoring.error",
  "profile.scoring.update",
  "profile.revealing.success",
  "profile.revealing.error",
  "profile.revealing.update",
  "profile.reasoning.success",
  "profile.reasoning.error",
  "profile.reasoning.update",
  "profile.embedding.success",
  "profile.embedding.error",
  "profile.embedding.update",
  "job.parsing.success",
  "job.parsing.error",
  "job.parsing.update",
  "job.scoring.success",
  "job.scoring.error",
  "job.scoring.update",
  "job.revealing.success",
  "job.revealing.error",
  "job.revealing.update",
  "job.reasoning.success",
  "job.reasoning.error",
  "job.reasoning.update",
  "job.embedding.success",
  "job.embedding.error",
  "job.embedding.update",
  "action.stage.success",
  "action.stage.error",
  "action.rating.success",
  "action.rating.error",
  "agent.training.success",
  "agent.training.error",
  "agent.training.update",
  "agent.scoring.success",
  "agent.scoring.error",
  "agent.scoring.update",

];