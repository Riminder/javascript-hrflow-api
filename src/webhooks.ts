import Hrflow = require("./index");
import Events from "./events";
import * as util from "tweetnacl-util";
import * as sha256 from "fast-sha256";
import { WebhooksResponse } from "./types";
import defaults from "./defaults";
import { httpPostRequest } from "./http";

export namespace Webhooks {
  export interface ResponseBase {
    type: string;
    message: string;
  }
  export interface ProfileParseResponse extends ResponseBase {
    profile: Profile;
  }

  export interface ProfileScoreResponse extends ResponseBase {
    profile: Profile;
    filter: Filter;
    score: number;
  }

  export interface FilterTrainResponse extends ResponseBase {
    filter: Filter;
  }

  export interface FilterScoreResponse extends ResponseBase {
    filter: Filter;
  }

  export interface ActionStageResponse extends ResponseBase {
    profile: Profile;
    filter: Filter;
    stage: string;
  }

  export interface ActionRatingResponse extends ResponseBase {
    profile: Profile;
    filter: Filter;
    rating: number;
  }

  export type Response = ProfileParseResponse |
    ProfileScoreResponse |
    FilterTrainResponse |
    FilterScoreResponse |
    ActionRatingResponse |
    ActionStageResponse;

  export interface Profile {
    profile_key: string;
    profile_reference: string;
  }

  export interface Filter {
    filter_id: string;
    filter_reference: string;
  }

  export type EventCallbackMap = Map<string, (data: Webhooks.Response, type: string) => any>;
}

export class Webhooks {
  private hrflow: Hrflow;
  binding: Webhooks.EventCallbackMap;

  constructor(hrflow: Hrflow) {
    if (!hrflow  || !hrflow.Webhooks_Key) {
      throw new Error("The webhook secret key must be specified");
    }

    this.hrflow = hrflow;
    this.binding = new Map<string, (data: Webhooks.Response, type: string) => any>();
  }

  handle(headers: any): () => void {
    return () => {
      if (!headers["HTTP-RIMINDER-SIGNATURE"]) {
        throw new Error("The signature is missing from the headers");
      }

      const [encodedSignature, encodedPayload] = headers["HTTP-RIMINDER-SIGNATURE"].split(".");
      const expectedSignature = util.encodeBase64(sha256.hmac(util.decodeUTF8(this.hrflow.Webhooks_Key), util.decodeUTF8(encodedPayload)));

      if (encodedSignature !== expectedSignature) {
        throw new Error("The signature is invalid");
      }

      const payload: Webhooks.Response = JSON.parse(util.encodeUTF8(util.decodeBase64(encodedPayload)));

      if (Events.indexOf(payload.type) < 0) {
        throw new Error(`Unknown event: ${payload.type}`);
      }
      this._callBinding(payload);
    };
  }

  on(event: string, callback: (data: Webhooks.Response, type?: string) => any) {
    if (Events.indexOf(event) < 0) {
      throw new Error("This event doesn't exist");
    }

    if (this.binding.has(event)) {
      throw new Error("This callback already has been declared");
    }

    this.binding.set(event, callback);

    return this;
  }

  check(): Promise<WebhooksResponse> {
    const url = `${defaults.API_URL}/webhook/check`;
    return httpPostRequest(url, null, { headers: this.hrflow.headers });
  }

  private _callBinding(payload: Webhooks.Response): void {
    if (this.binding.has(payload.type)) {
      this.binding.get(payload.type)(payload, payload.type);
    }
  }
}