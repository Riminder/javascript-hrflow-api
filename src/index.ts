import { HrflowOptions } from "./types";
import { Webhooks } from "./webhooks";
import Source from "./resources/source";
import Profile from "./resources/profile/profile";
import Offer from "./resources/job/job";

class Hrflow {
  public headers: any;
  public API_Key: string;
  public Webhooks_Key: string;
  public webhooks: Webhooks;
  public source: Source;
  public profile: Profile;
  public offer: Offer;

  constructor(options: HrflowOptions) {

    if (!options.API_Key) {
      let error = new Error("No API Key was supplied for Hrflow SDK");
      throw error;
    }

    this.API_Key = options.API_Key;
    this.headers = {
      "X-API-Key": this.API_Key,
    };
    if (options.Webhooks_Key) {
      this.Webhooks_Key = options.Webhooks_Key;
    }

    this._init();
  }

  private _init() {
    if (this.Webhooks_Key) {
      this.webhooks = new Webhooks(this);
    }

    this.source = new Source(this);
    this.profile = new Profile(this);
    this.offer = new Offer(this);
  }
}

export = Hrflow;