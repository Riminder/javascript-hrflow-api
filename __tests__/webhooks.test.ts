import Hrflow = require("../src/index");
import { Webhooks } from "../src/webhooks";
import * as util from "tweetnacl-util";
import * as sha256 from "fast-sha256";

let app: Hrflow;
const api_secret = "api_key";
const api_user = "api_user";
const webhooks_key = "webhooks_key";

function generateSignature(key: string, event: string) {
    const data: Webhooks.ProfileParseResponse = {
        type: event,
        message: "The parsing succeeded",
        profile: {
            profile_key: "id",
            reference: "reference"
        }
    };

    const JSONPayload = JSON.stringify(data);
    const base64Payload = util.encodeBase64(util.decodeUTF8(JSONPayload));

    const signature = sha256.hmac(util.decodeUTF8(key), util.decodeUTF8(base64Payload));
    const base64Signature = util.encodeBase64(signature);

    return `${base64Signature}.${base64Payload}`;
}

const callbackMock = jest.fn();

describe("Webhooks tests",  () => {
    describe("Webhook check", () => {
        test("It should check if the webhook is correctly set up", () => {
            app = new Hrflow({ api_secret, api_user, webhooks_key });
            app.webhooks.check()
                .then((response) => {
                    expect(response).toMatchSnapshot();
                });
        });
    });

    describe("Webhook creation", () => {
        test("It should not create webhooks if no key is given", () => {
            app = new Hrflow({ api_secret, api_user });
            expect(app.webhooks).toBeUndefined();
        });

        test("It should create webhooks if the key is given", () => {
            app = new Hrflow({ api_secret, api_user, webhooks_key });
            expect(app.webhooks).toBeDefined();
        });

        test("It should throw an error if we try to create webhooks without a key", () => {
            expect(() => {
                const webhooks = new Webhooks(null);
            }).toThrowError("The webhook secret key must be specified");
        });
    });

    describe("Webhooks usage", () => {
        describe("Webhooks binding", () => {
            beforeEach(() => {
                app = new Hrflow({ api_secret, api_user, webhooks_key });
            });

            test("It should throw an error if we try to bind an inexistant event", (done) => {
                expect(() => {
                    app.webhooks.on("not.an.event", (data: Webhooks.Response, type: string) => {});
                }).toThrowError("This event doesn't exist");
                done();
            });

            test("It should throw an error if we try to bind two functions to the same event", (done) => {
                expect(() => {
                    app.webhooks
                        .on("profile.parsing.success", (data: Webhooks.Response, type: string) => {})
                        .on("profile.parsing.success", (data: Webhooks.Response, type: string) => {});
                }).toThrowError("This callback already has been declared");
                done();
            });

            test("It should bind the event correctly", (done) => {
                expect(app.webhooks.on("profile.parsing.success", (data: Webhooks.Response, type: string) => {
                    console.log(type);
                    return 42;
                })).toBeInstanceOf(Webhooks);
                expect(app.webhooks.binding.has("profile.parsing.success")).toBeTruthy();
                expect(app.webhooks.binding.get("profile.parsing.success")({} as Webhooks.Response, "test")).toBe(42);
                done();
            });
        });

        describe("Webhooks call", () => {
            beforeAll(() => {
                app = new Hrflow({ api_secret, api_user, webhooks_key });
                app.webhooks.on("profile.parsing.success", callbackMock);
            });

            test("It should throw an error if the header is not given", () => {
                expect(app.webhooks.handle({})).toThrowError("The signature is missing from the headers");
            });

            test("It should throw an error if the signature is invalid", () => {
                const signature = generateSignature("wrong_key", "profile.parsing.success");
                const headers = {
                    "HTTP-RIMINDER-SIGNATURE": signature
                };
                expect(app.webhooks.handle(headers)).toThrowError("The signature is invalid");
            });

            test("It should throw an error if the event is unknown", () => {
                const signature = generateSignature(webhooks_key, "unknwown.event");
                const headers = {
                    "HTTP-RIMINDER-SIGNATURE": signature
                };
                expect(app.webhooks.handle(headers)).toThrowError("Unknown event: unknwown.event");
            });

            test("It should call the callbacj function", () => {
                const signature = generateSignature(webhooks_key, "profile.parsing.error");
                const headers = {
                    "HTTP-RIMINDER-SIGNATURE": signature
                };
                app.webhooks.on("profile.parsing.error", callbackMock);
                app.webhooks.handle(headers)();
                expect(callbackMock).toHaveBeenCalledTimes(1);
            });
        });
    });
});