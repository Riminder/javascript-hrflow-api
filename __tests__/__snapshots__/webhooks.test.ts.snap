// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webhooks tests Webhook check It should check if the webhook is correctly set up 1`] = `
Object {
  "options": Object {
    "body": null,
    "headers": Object {
      "X-API-KEY": "api_key",
      "X-USER-EMAIL": undefined,
    },
    "method": "POST",
  },
  "url": Url {
    "auth": null,
    "hash": null,
    "host": "api.hrflow.ai",
    "hostname": "api.hrflow.ai",
    "href": "https://api.hrflow.ai/v1/webhook/check",
    "path": "/v1/webhook/check",
    "pathname": "/v1/webhook/check",
    "port": null,
    "protocol": "https:",
    "query": null,
    "search": null,
    "slashes": true,
  },
}
`;
